### ISSUE TITLE
quotes single do not trigger warning when backtick quotes are used

### ISSUE THREAD

**Issue by gajus** [eslint/eslint#4260]
```json
"quotes": [
    2,
    "single"
],
```

Using the above configuration, no error is produced for the following code:

```js
import {
    expect
} from 'chai';

import _ from 'lodash';

import table from './../../../src';

describe(`README.md usage/`, () => {
    let expectTable;

    before(() => {
        expectTable = (rows, contents) => {
            expect(rows).to.equal(_.trim(contents) + `\n`);
        };
    });

    it(`text_truncating`, () => {
        let config,
            data,
            output;

        data = [
            ['Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus pulvinar nibh sed mauris convallis dapibus. Nunc venenatis tempus nulla sit amet viverra.']
        ];

        config = {
            columns: {
                0: {
                    width: 20,
                    truncate: 100
                }
            }
        };

        output = table(data, config);

        // console.log(output);

        expectTable(output, `
╔══════════════════════╗
║ Lorem ipsum dolor si ║
║ t amet, consectetur  ║
║ adipiscing elit. Pha ║
║ sellus pulvinar nibh ║
║ sed mauris conva...  ║
╚══════════════════════╝
        `);
    });
});
```

**Comment by gajus** [comment]
Using 1.7.3.

**Comment by nzakas** [comment]
Backticks are allowed when you specify "single" or "double", this is intentional.

**Comment by sindresorhus** [comment]
Potential new rule idea: Only allow template literals when they're actually useful, when normal quotes can't be used: Multiline, interpolation, contains same quote character as wrapper.

**Comment by gajus** [comment]
Where is this documented?

**Comment by gajus** [comment]
What is the logic for this?

I think it should work as follows:

* When "quotes" is to "single" only allow single quotes.
* Have a "allow-backtick-when-multiline" flag that allows to use backtick quotes when a string spans multiple lines.
* Have a "allow-backtick-when-variable" flag that allows to use backtick quotes when string includes a variable.

**Comment by ilyavolodin** [comment]
Backticks are superset of quotes in ES6. Having them flagged means that you are disabling one of the language features, since there is no way to use templates with quotes.

**Comment by gajus** [comment]
> Backticks are superset of quotes in ES6.

What does that mean?

From linter perspective, ES5 template strings is another type of quotes. In similar way that ESLint now allows to restrict quotes to double or single, it should restrict them to either double, single or backtick.

**Comment by nzakas** [comment]
It's not obvious in the docs, but if you look at the examples, there's this:

```
/*eslint quotes: [2, "double"]*/
/*eslint-env es6*/

var double = "double";
var backtick = `backtick`; // backticks are allowed
```

We can definitely update the docs to be more explicit.

The rationale is while people will undoubtedly be opinionated about single vs. double quotes, it's unlikely people will want to flat-out disable templates just because they use single or double quotes for strings. It's also unlikely that most people will choose to use templates only and disable single and double quotes altogether. It's a gray area, for sure.

**Comment by gajus** [comment]
So, for now, if I want to flat out disable template strings, I need to write a custom plugin

**Comment by lo1tuma** [comment]
@gajus If you are using the default parse you could simple disable the [language option](http://eslint.org/docs/user-guide/configuring#specifying-language-options) `templateStrings`.

**Comment by IanVS** [comment]
> Only allow template literals when they're actually useful, when normal quotes can't be used
@sindresorhus

I'm :+1: on this.  The logic could even be reused from [prefer-template](http://eslint.org/docs/rules/prefer-template), but instead of warning when templates are not used where they could add value, `quotes` would warn when they are used in cases where they don't add real value.

**Comment by gajus** [comment]
> @gajus If you are using the default parser you could simply disable the language option templateStrings.

I am using babel-eslint.

**Comment by gajus** [comment]
I don't see any potential drawback of enforcing a particular quote type and using flags to optionally allow template tags in certain scenarios (as proposed https://github.com/eslint/eslint/issues/4260#issuecomment-151223214).

What are the arguments against it?

**Comment by IanVS** [comment]
Generally we try to think long and hard before adding options, as each one increases the cognitive load for users, because they need to understand it and then decide whether they should set it, and if so, how.  In this case, it would  introduce a breaking change, which we also try to minimize.  I can't imagine there are too many people who would say "I'm okay with backticks for multiline, but not for interpolation" (which would basically banish use of interpolation).  I'm of the opinion that it makes sense to prohibit backticks in cases where they could be directly replaced by the string type for which the option is set, and allow them in all other situations.

**Comment by mysticatea** [comment]
Related: https://github.com/eslint/eslint/issues/3090

**Comment by gajus** [comment]
Anyone wanting to write a simple ESLint plugin to restrict use of template strings?

**Comment by IanVS** [comment]
Thanks @mysticatea, I thought this discussion seemed familiar. Looks like it was left hanging though.

**Comment by gajus** [comment]
It is not the same. I want to completely restrict use of template strings when "single" quotes rule is set.

**Comment by mysticatea** [comment]
@gajus Though this may not be the best way, I guessed you can use [no-restricted-syntax](http://eslint.org/docs/rules/no-restricted-syntax) rule for this.

For Example:

```js
/* eslint no-restricted-syntax: [2, "TemplateLiteral"] */

let foo = `hello, ${name}!`;  // Using "TemplateLiteral" is not allowed.
```

**Comment by michaelficarra** [comment]
@mysticatea That looks like a good solution to @gajus's problem.

@sindresorhus Please voice your support in #3090.

**Comment by gajus** [comment]
@mysticatea Thank you.

**Comment by gajus** [comment]
For me this solution is good enough. Happy to close this issue. Though, will leave it up to the team to do it in case anyone have any other concerns.

**Comment by nzakas** [comment]
Let's regroup on #3090. Thanks all.
